### PR TITLE
Problem: no Byzantine integration test(fix #365) 

### LINKED ISSUE #365: Problem: no Byzantine integration test
I guess could be similar to https://github.com/crypto-com/thaler/blob/main/integration-tests/multinode/byzantine_test.py
after the bech32 fix is merged

### PR DESCRIPTION
**Pull request by linfeng-crypto**
- create 3 validator nodes, keep node0 has more than  2/3 voting power
- stop node2, and copy node1's validator key file to node2
- start node2, waiting for node1/node2 jailed
- check the tokens after node slashed and byzantine error

### CODE REVIEW

**Comment by lezzokafka** [comment]
Should we also include the slashing part in this integration test?

**Review by tomtau — CHANGES_REQUESTED** [review]
> Should we also include the slashing part in this integration test?

yeah, will be good to add test assertions that the account balance was decreased / slashed

**Comment by tomtau** [comment]
I guess the integration test fail isn't relevant to this PR: https://github.com/crypto-org-chain/chain-main/pull/410/checks?check_run_id=2082753021#step:5:2026 ? is it https://github.com/crypto-org-chain/chain-main/issues/385 @yihuang  ?

**Review by yihuang — APPROVED** [review]
LGTM even better if we can assert the reason of jailing

**Comment by yihuang** [comment]
> I guess the integration test fail isn't relevant to this PR: https://github.com/crypto-org-chain/chain-main/pull/410/checks?check_run_id=2082753021#step:5:2026 ? is it #385 @yihuang ?

yeah, i think so.

**Comment by tomtau** [comment]
there are two small lint issues in the Python script, but otherwise good to be merged

**Review by devashishdxt — APPROVED** [review]

**Comment by tomtau** [comment]
@yihuang it seems the IBC test consistently fails on this PR -- due to the address being in use: https://github.com/crypto-org-chain/chain-main/pull/410/checks?check_run_id=2083323245#step:5:997

is it that the node doesn't get stopped here?

**Comment by linfeng-crypto** [comment]
> LGTM even better if we can assert the reason of jailing

I tried, but can't the details jailed reason, it seems that there is no API to get the reason.

File: integration_tests/test_byzantine.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    # it may take 30s to finish the loop
+    i = 0
+    while i < 300:
+        time.sleep(1)
+        sys.stdout.write(".")
+        sys.stdout.flush()
+        i += 1
+        val_addr = cluster.address("validator", from_node, bech="val")
+        val_1 = cluster.validator(val_addr)
```
Comment (leejw51crypto) [inline integration_tests/test_byzantine.py:51]: how about val_1 -> validator1

File: integration_tests/test_byzantine.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        if val_1["jailed"]:
+            break
+    assert val_1["jailed"]
+    assert val_1["status"] == "BOND_STATUS_UNBONDING"
+    print("\n{}s waiting for node 1 jailed".format(i))
+
+    i = 0
+    # it may take 2min to finish the loop
+    while i < 300:
```
Comment (leejw51crypto) [inline integration_tests/test_byzantine.py:60]: declare as const , in move to the top of source code?

**Review by leejw51crypto — APPROVED** [review]
lgtm,
small naming suggestions

**Comment by JayT106** [comment]
LGTM, just thinking could we just use one while loop to check the status on both nodes? Does the consensus behavior always jail the node1 first and then jail the node2?

**Review by tomtau — CHANGES_REQUESTED** [review]
it seems "test_ibc" consistently fails on this PR: https://github.com/crypto-org-chain/chain-main/pull/410/checks?check_run_id=2086193872#step:5:2085
```
ibc-0/node0.log 12:42PM INF Starting Node service impl=Node
1509
ibc-0/node0.log Error: failed to listen on 0.0.0.0:26657: listen tcp 0.0.0.0:26657: bind: address already in use
1510
ibc-0/node0.log Usage:
1511
ibc-0/node0.log   chain-maind start [flags]
1512
ibc-0/node0.log 
...
```

so either disable/ignore that test for now or fix it -- @yihuang may help?

**Comment by yihuang** [comment]
> it seems "test_ibc" consistently fails on this PR: https://github.com/crypto-org-chain/chain-main/pull/410/checks?check_run_id=2086193872#step:5:2085
> 
> ```
> ibc-0/node0.log 12:42PM INF Starting Node service impl=Node
> 1509
> ibc-0/node0.log Error: failed to listen on 0.0.0.0:26657: listen tcp 0.0.0.0:26657: bind: address already in use
> 1510
> ibc-0/node0.log Usage:
> 1511
> ibc-0/node0.log   chain-maind start [flags]
> 1512
> ibc-0/node0.log 
> ...
> ```
> 
> so either disable/ignore that test for now or fix it -- @yihuang may help?

relayer execute failed:
```
E               subprocess.CalledProcessError: Command '['relayer', '--home', PosixPath('/tmp/pytest-of-runner/pytest-0/popen-gw0/data1/relayer'), 'light', 'init', 'ibc-0', '-f']' returned non-zero exit status 1.
```

not sure why.

File: pystarport/pystarport/cluster.py [hunk 3]
```diff
@@ -698,7 +740,7 @@ def init_cluster(
                 config.keys(), list(rly_section.get("paths", {}).keys())
             ),
         )
-    if len(chains) > 1:
+    if config.get("relayer"):
```
Comment (yihuang) [inline pystarport/pystarport/cluster.py:743]: this change maybe the closest to the relayer issue in integration tests, can try to revert it and try again.

File: integration_tests/test_byzantine.py (hunk 1 above)
Comment (linfeng-crypto) [reply integration_tests/test_byzantine.py:51]: Ok

File: integration_tests/test_byzantine.py (hunk 2 above)
Comment (linfeng-crypto) [reply integration_tests/test_byzantine.py:60]: Ok

File: pystarport/pystarport/cluster.py (hunk 3 above)
Comment (linfeng-crypto) [reply pystarport/pystarport/cluster.py:743]: Ok

**Comment by tomtau** [comment]
still https://github.com/crypto-org-chain/chain-main/runs/2092934129#step:5:2049 🥺

File: integration_tests/configs/byzantine.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+chain_id_1:
+  validators:
+    - coins: 8000cro
+      staked: 8000cro
+      base_port: 26650
```
Comment (yihuang) [inline integration_tests/configs/byzantine.yaml:5]: I guess this is the issue, don't set port explicitly, it'll be generated automatically.

Comment (tomtau) [reply integration_tests/configs/byzantine.yaml:5]: @linfeng-crypto

File: integration_tests/test_j_byzantine.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import sys
```
Comment (yihuang) [inline integration_tests/test_j_byzantine.py:1]: is there a typo in the filename?

**Review by tomtau — APPROVED** [review]

**Pull request merged**